vpr/src/pack/greedy_candidate_selector.cpp

```diff
@@ -0,0 +1,1076 @@
+        cluster_gain_stats.num_pins_of_net_in_pb[net_id] = 0;
+    }
+    cluster_gain_stats.num_pins_of_net_in_pb[net_id]++;
+}
+
+/*
+ * @brief Determine if atom block is in pb.
+ *
+ * TODO: This would make more sense in the cluster legalizer class.
+ */
+static bool is_atom_blk_in_pb(const AtomBlockId blk_id, const t_pb* pb) {
+    const AtomContext& atom_ctx = g_vpr_ctx.atom();
+
+    const t_pb* cur_pb = atom_ctx.lookup.atom_pb(blk_id);
+    while (cur_pb) {
+        if (cur_pb == pb) {
+            return true;
+        }
+        cur_pb = cur_pb->parent_pb;
+    }
+    return false;
+}
+
+void GreedyCandidateSelector::update_connection_gain_values(
+                                       ClusterGainStats& cluster_gain_stats,
+                                       AtomNetId net_id,
+                                       AtomBlockId clustered_blk_id,
+                                       const ClusterLegalizer& cluster_legalizer,
+                                       e_net_relation_to_clustered_block net_relation_to_clustered_block) {
+
+    /*This function is called when the connectiongain values on the net net_id*
+     *require updating.   */
+
+    // Atom Context used to lookup the atom pb.
+    // TODO: Should investigate this. Using the atom pb in this class is very
+    //       strange.
+    const AtomContext& atom_ctx = g_vpr_ctx.atom();
+
+    int num_internal_connections, num_open_connections, num_stuck_connections;
+    num_internal_connections = num_open_connections = num_stuck_connections = 0;
+
+    LegalizationClusterId legalization_cluster_id = cluster_legalizer.get_atom_cluster(clustered_blk_id);
+
+    /* may wish to speed things up by ignoring clock nets since they are high fanout */
+    for (AtomPinId pin_id : atom_netlist_.net_pins(net_id)) {
+        AtomBlockId blk_id = atom_netlist_.pin_block(pin_id);
+        if (cluster_legalizer.get_atom_cluster(blk_id) == legalization_cluster_id
+            && is_atom_blk_in_pb(blk_id, atom_ctx.lookup.atom_pb(clustered_blk_id))) {
+            num_internal_connections++;
+        } else if (!cluster_legalizer.is_atom_clustered(blk_id)) {
+            num_open_connections++;
+        } else {
+            num_stuck_connections++;
+        }
+    }
+
+    if (net_relation_to_clustered_block == e_net_relation_to_clustered_block::OUTPUT) {
+        for (AtomPinId pin_id : atom_netlist_.net_sinks(net_id)) {
+            AtomBlockId blk_id = atom_netlist_.pin_block(pin_id);
+            VTR_ASSERT(blk_id);
+
+            if (!cluster_legalizer.is_atom_clustered(blk_id)) {
+                /* TODO: Gain function accurate only if net has one connection to block,
+                 * TODO: Should we handle case where net has multi-connection to block?
+                 *       Gain computation is only off by a bit in this case */
+                if (cluster_gain_stats.connectiongain.count(blk_id) == 0) {
+                    cluster_gain_stats.connectiongain[blk_id] = 0;
```

Comment:
I thought most of this became connection_gain --> do a check and see if you can make the naming consistent (or leave the comments alone).